### PR TITLE
!!! [v6r12] FixRequest optimization + DM for aliases

### PR DESCRIPTION
**Pull request by phicharp**
This PR makes an important fix:

* Requests optimization was buggy in case the failover operations sequence was not first in the request: it was putting the merged operations upfront. This was making requests fail...

* In addition, request made for checking the presence of a file not only in the destination SE but also in aliases (DataManager)

### CODE REVIEW

File: DataManagementSystem/Client/DataManager.py [hunk 1]
```diff
@@ -582,6 +582,37 @@ def replicate( self, lfn, destSE, sourceSE = '', destPath = '', localCache = ''
       return res
     return S_OK( lfn )
 
+  def __getSERealName( self, storageName ):
```
Comment (chaen) [inline DataManagementSystem/Client/DataManager.py]: This should be a method in a utility, not in DM class

File: RequestManagementSystem/Client/Request.py [hunk 2]
```diff
@@ -534,10 +534,13 @@ def optimize( self ):
     repAndRegList = []
     removeRepList = []
     i = 0
-    while i < len( self.__operations__ ) - 1:
```
Comment (chaen) [inline RequestManagementSystem/Client/Request.py]: Isn't it doing the exact same ?

File: RequestManagementSystem/Client/Request.py [hunk 3]
```diff
@@ -550,22 +553,34 @@ def optimize( self ):
           removeRepList.append( ( op2.TargetSE, op2 ) )
           del self.__operations__[i]
           del self.__operations__[i]
-          continue
+          print 'Found sequence starting at %d' % i
```
Comment (chaen) [inline RequestManagementSystem/Client/Request.py]: print to be replaced with gLogger

File: Resources/Storage/StorageFactory.py [hunk 4]
```diff
@@ -11,6 +11,7 @@
     getStorages()      This takes a DIRAC SE definition and creates storage stubs for the protocols found in the CS.
                       By providing an optional list of protocols it is possible to limit the created stubs.
 """
+from __builtin__ import False
```
Comment (chaen) [inline Resources/Storage/StorageFactory.py]: I think Eclipse added this for free :-)

File: RequestManagementSystem/Client/Request.py [hunk 5]
```diff
@@ -550,22 +553,34 @@ def optimize( self ):
...
+          optFound = True
+      if not optFound:
```
Comment (chaen) [inline RequestManagementSystem/Client/Request.py]: I am afraid the algorithm doesn't work in case your Request is only made of the 2 Operations ReplicateAndRegister and RemoveReplica.
First loop, i =0, you put the 2 operations in your lists, and remove them from the operation list. Since optFound is True, we don't increment i, and go back to the beginning of the loop. 
Then at the start of the second loop, i is still 0, ben len(self.__operations__) is 0, so you exit the loop, without ever inserting back the operations.
I just tried it on a toy example, and it is indeed the case.

May I recommend you to write unit test for this method ? No mock required, and VERY useful for future evolution ;-) I can even help there if you want

**Comment by atsareg** [comment]
Please, reconsider the last review comments and confirm if the patch is to be reworked or can be incorporated

**Comment by phicharp** [comment]
Yes, Chris is perfectly right... It fixes the problem we outlined but doesn't work in the most common case of only 2 operations... Stupid me as I only tested it fixes the problem :(
I'll fix the code now...

File: DataManagementSystem/Client/DataManager.py (hunk 1 above)
Comment (phicharp) [inline DataManagementSystem/Client/DataManager.py]: Well, this is questionable: if a method is only called at one place, why create an artificial dependency? I agree it can be a function and not a method... Will do the change...

File: RequestManagementSystem/Client/Request.py (hunk 2 above)
Comment (phicharp) [inline RequestManagementSystem/Client/Request.py]: I can't remember why I did that but there must have been a good reason... will comment when fixing the code ;-)

File: RequestManagementSystem/Client/Request.py (hunk 3 above)
Comment (phicharp) [inline RequestManagementSystem/Client/Request.py]: Oups, debug printout, sorry...

File: RequestManagementSystem/Client/Request.py (hunk 5 above)
Comment (phicharp) [inline RequestManagementSystem/Client/Request.py]: Perfectly correct... It fixes the problem but creates an even bigger one! 
Conclusion: don't implement a fix while listening to a conference!

File: Resources/Storage/StorageFactory.py (hunk 4 above)
Comment (phicharp) [inline Resources/Storage/StorageFactory.py]: Yes, Eclipse is now dong very strange things if one's not careful like adding unwanted lines or strings... Not clear if this is a bad setting or effect of a new version

File: RequestManagementSystem/Client/Request.py (hunk 2 above)
Comment (phicharp) [inline RequestManagementSystem/Client/Request.py]: Ah, yes! I was scared that since the number of operations is changing within the loop the length would not be reevaluated at each loop (kind of optimisation)... I now checked with an example and this is not the case. len() is reevaluated at each loop... Changed back...

**Comment by phicharp** [comment]
@Chris: can you have a look at it now... It was tested with the unit test, and therefore should be OK

File: Resources/Storage/StorageFactory.py (hunk 4 above)
Comment (chaen) [inline Resources/Storage/StorageFactory.py]: Don't forget to remove that

**Comment by chaen** [comment]
Besides the 'import False' which is still there, if the unit test says okay, I'm fine with it as well :-)

**Comment by phicharp** [comment]
Removed the spurious import ! ;-)

**Comment by atsareg** [comment]
I may be missed some discussion but: what is the difference between the Alias and BaseSE ? What is the reason to introduce a new concept ? This important new change is not commented in the PR

**Comment by phicharp** [comment]
The difference is quite simple:

* An alias is just a reference: if you register a file ALIAS-SE that is an alias to MY-SE, the file is registered in the FC as MY-SE. It is mostly useful when files have been registered as ALIAS-SE and one wants to get rid of this name just by disappearance of replicas

* A BaseSE is a real SE, but its properties are derived from another SE. The idea behind that is to avoid having multiple identical SEs except the name (e.g. CNAF-DST is totally identical to CNAF_M-DST, CNAF_MC-DST etc... Using a base SE allows to have the definition appearing only once, but the actual name to be used for registration and accounting.

In other words, Alias is a reference, BaseSE is an inheritance.

File: Resources/Storage/StorageFactory.py [hunk 6]
```diff
@@ -203,9 +210,34 @@ def _getConfigStorageName( self, storageName ):
...
+      errStr = "StorageFactory._getConfigStorageName: Supplied storage doesn't exist."
+      gLogger.error( errStr, configPath )
+      return S_ERROR( errStr )
+    if 'BaseSE' in res['Value']:
+      configPath = '%s/%s/BaseSE' % ( self.rootConfigPath, storageName )
+      baseName = gConfig.getValue( configPath )
+      # Just in case the base is an alias ;-)
+      result = self._getConfigStorageName( baseName )
+      if not result['OK']:
+        return result
```
Comment (atsareg) [inline Resources/Storage/StorageFactory.py]: This new function is just a duplication of code, can be made a single function with an extra argument, e.g. referenceType, that can take "Alias" and "BaseSE" string values. If BaseSE concept is introduced, then eventually the SE definition can take all values from the Base SE and be able to override some of them in its own definition

File: RequestManagementSystem/Client/Request.py [hunk 7]
```diff
@@ -26,7 +26,7 @@
 # # imports
 import datetime
 # # from DIRAC
-from DIRAC import S_OK, S_ERROR
+from DIRAC import S_OK, S_ERROR, gLogger
```
Comment (atsareg) [inline RequestManagementSystem/Client/Request.py]: There is no use of gLogger in this module
